feat(article): add getting started with tRPC article (#111)
iodigital-com/io-technology: data/blog/getting-started-with-trpc/part-1-building-the-api.md

---

> title: 'Getting started with tRPC v10 by building a todo app - part 1'

Reviewer: The website will show that it is a serie so can you remove the " - part 1" from the title?

> date: '2022-10-10'

Reviewer 2: Change date right before you merge 😉

tags: ['frontend', 'react', 'trpc', 'react native']
images: ['/articles/getting-started-with-trpc/hero.png']
summary: "There's nothing quite like the feeling of finally getting your front-end and back-end types to match up. After hours of slacking, emailing, and going through documentation, you finally have it! But that feeling doesn't last long, because you realize the back-end had some type changes, which broke your front-end again! Well, here is where tRPC comes into play."
authors: ['zjerlondy-ferero']
theme: 'rouge'
serie: 'getting-started-with-trpc'
---

> # Getting started with tRPC v10 by building a todo app part 1

Reviewer: The h1 is already added to the page because of the frontmatter. Can you remove this one?

[tRPC](https://trpc.io/) is a lightweight library that allows you to create fully typesafe APIs without schemas or code generation. It provides end-to-end typing, which means that types can be shared between the server and client(s). This results in fewer bugs and less boilerplate code.

> In this two-part series, we'll be building a todo app using tRPC. In the first part, we'll start by creating a simple backend API using tRPC, and then in the second part, we'll create a (React Native) frontend UI that consumes the API. By the end of this series, you'll have a fully functioning todo app!

Reviewer 2: I would remove "simple" from "simple backend". It's making an assumption but you don't know the knowledge level of the audience.

![what we are creating](/articles/getting-started-with-trpc/what-we-are-creating.png)

## Prerequisites

> - [Node](https://nodejs.org/en/) installed

Reviewer 2: Is there a minimal version supported? Maybe add it if so

- [React (Native)](https://reactnative.dev/docs/environment-setup) development environment
- Basic React (Native) / Typescript knowledge

We won't be focusing on installing and setting up Typescript in this article; therefore, I've created a [Github repo](https://github.com/LondyF/getting-started-with-trpc) with some basic Typescript boilerplate to get us started right away!

## Setting up our Express/tRPC backend

> In this article, we will be using tRPC in conjunction with [ExpressJS](https://expressjs.com/). tRPC has a nice ExpressJS adapter that handles some of the tRPC magic for us, thus making setting up our tRPC backend very easy. Even though we use ExpressJS, tRPC can be used in combination with any (Node) backend that supports Typescript.

Reviewer 2: Add an extra paragraph here with mentioning tRPC is often used in combination with Next.js but in this article you are using ExpressJS to also show that usecase.

### Installing dependecies

> Let's start by using the terminal to navigate to the `server` directory in our project directory, and install the following dependecies:

Reviewer 2: You could add the `cd` command from the root of your supplied getting started repo

`yarn add @trpc/server@next zod`

> As you can see, we install the `@trpc/server` package, but also a package named `Zod`. [Zod](https://github.com/colinhacks/zod) is a library that makes input validation very simple by using schema-based validation. With Zod, tRPC can validate incoming requests against a predefined schema, keeping our handler function clean and free of any unnecessary validation checks. tRPC has out-of-the-box support for different scheme validation tools such as [Yup](https://github.com/jquense/yup) and [Superstruct](https://github.com/ianstormtaylor/superstruct).

Reviewer 2: "scheme validation tools"? or "schema validation tools"?

### Initializing tRPC and creating our first router

Let's start by creating a new file called `trpc.ts` in the `server/src/` of our project folder. In the file, add the following code:

```ts
import { initTRPC } from '@trpc/server'

> export const t = initTRPC.create()

Reviewer 2: imo `t` is quite short. Should you not do `trpc`?

```

All this file is responsible for is initializing and exporting a tRPC instance.

Now, let's start by creating, arguably, the most important piece of our back-end, the router. Begin by creating a new file called `todo.ts` in the `server/src/routers/` folder of our project folder. In the `todo.ts` add the following code:

```ts
// Import our tRPC instance
import { t } from '../trpc'

// Initalize an empty array where we will be storing our todo's.
// For now, we will use the type `any[]`, but this will be changed later on.
let todos: any[] = []

> // Create our todo router, and add a query (equivalent of a REST Get request) procedure called `all`,

Reviewer 2: ```suggestion
// Create our todo router, and add a query procedure (equivalent of a REST Get request) called `all`,
```

// which will be responsible for returning all the stored todo's
export const todoRouter = t.router({
  all: t.procedure.query(() => {
    return todos
  }),
})
```

> We create a tRPC router by calling the `router()` method and passing an object containing the different endpoints and their procedure as an argument. tRPC knows two procedures:

Reviewer 2: can procedure be plural?

import { t } from '../trpc'

// Import our todo Router
import { todoRouter } from './todo'

// Create an appRouter which will be used to tie together all our different routers
// In our case, we will only have one router, our todo router. This todo router will be exported under the namespace `todo`.
export const appRouter = t.router({
  todo: todoRouter,
})

// Export only the **type** of a router to avoid importing server code on the client
export type AppRouter = typeof appRouter
```

As you can see, we start again by importing our tRPC instance, followed by importing our `todoRouter`. After that, we create a new `appRouter` which will be used to tie together all our different routers. In our case, this will only be our `todoRouter`, which is exported under the namespace `todo`.

Namespaces are used to organize our routes. To consume our API in the front-end, you will make use of the namespaces to identify the correct router. For instance, if you wanted to query the todo router, you would write: `todo.` followed by the procedure name, in our case: `todo.all`.

Lastly, we export the type `AppRouter`, so that we can use it later on in the front-end.

### Creating our server

> Now that we have created our router, we would want to have a way to make calls to that router, so let's start by creating a simple Express server. In the `server/src` folder of our project, we will be creating a new file called `index.ts` which contains the following code:

Reviewer 2: "simple" 😉

// Initialize Express
const app = express()

// Tell Express to parse incoming requests using JSON
app.use(express.json())

// Tell Express to let the tRPC adapter handle all incoming requests to `/trpc`
app.use(
  '/trpc',
  trpcExpress.createExpressMiddleware({
    router: appRouter,
  })
)

// Start the server under the port 3000
app.listen(3000)
```

If you are familiar with Express, the code above is pretty self-explanatory. First, we start by importing the tRPC Express Adapter, Express and finally our `appRouter`. After that, we initialize Express, and tell Express to use the `express.json()` middleware to parse incoming requests to JSON. After that, we instruct Express to route all incoming calls to /trpc through a middleware exported from the trpcExpress.createExpressMiddleware() method. This method takes an option object as an argument. In our case, we only pass our `appRouter` as an option. So, all incoming requests to `/trpc` will be handled by our `appRouter`.

## Testing our first endpoint

Now that we have created our server and a simple route, it's time to test! We can do this by simply starting our server, hitting the API, and checking the response.

> Start by navigating to the project folder in the terminal and running `npm run dev`. This will start the server on port `3000`. Now, open a browser and navigate to `http://localhost:3000/trpc/todo.all`, which will result in the following JSON response:

Reviewer 2: Earlier in the article you've used `yarn` so you could do `yarn dev` here

  }
}
```

As you can see, it's adding `null` to our data array. This is happening due to us not specifying to tRPC what the body of our incoming request will look like, and therefore tRPC fails to parse the incoming request into usable data. There are several ways we can fix this, but we will fix this by simply declaring a specific input format for our `todo.add` procedure. Let's jump back to our todo router file, and modify our code to match the following code:

```ts
// Import our input validation tool
import { z } from 'zod'

// Import our tRPC instance
import { t } from '../trpc'

// Initalize an empty array where we will be storing our todo's.
// For now, we will use the type 'any[]', but this will be changed later on.
let todos: any[] = []

// Create our todo router, and add a query (equivalent of a REST Get request) procedure with the name `all`
// Which be responsible for returning all the stored todos
export const todoRouter = t.router({
  all: t.procedure.query(() => {
    return todos
  }),
  add: t.procedure
    .input(
      z.object({
        title: z.string(),
      })
    )
    .mutation(({ input }) => {
      todos.push(input)

      return todos
    }),
})
```

> If you look closely at our `add` procedure, you will notice we added a new method called `input()`. This `input()` method takes an input validation schema that defines what our incoming request body should look like. Like mentioned before, tRPC supports a handful of input validators, but we have decided to use `Zod`. If you look closely, you can see that all we are doing is telling tRPC that our request body should exist of an object containing a title that is a string. Let's jump back to Postman and hit our `todo.add` API again to test this out.

Reviewer 2: ```suggestion
If you look closely at the `add` procedure, you will notice we added a new method called `input()`. This `input()` method takes an input validation schema that defines what our incoming request body should look like. tRPC supports a handful of input validators, but we have decided to use `Zod`. If you look closely, you can see that all we are doing is telling tRPC that our request body should exist of an object containing a title that is a string. Let's jump back to Postman and hit our `todo.add` API again to test this out.
```

As you can see, the data array is now correctly populated:

```json
{
  "result": {
    "data": [
      {
        "title": "our first todo"
      }
    ]
  }
}
```

Now let's extend our request body object by adding the extra fields: `id` and `completed`. The `id` field will be used to uniquely identify a todo, and the `completed` field will be used to mark a todo as completed. To keep our `todoRouter` function clean and simple, I will be moving the add todo request body validation schema to a separate file. Let's create a new folder, `models`, in our server's `src/` directory. This folder will contain one file called `todo.ts`. This is where we will define our add todo validation schema.

```ts
/* server/src/models/todo.ts */

import { z } from "zod";

// Helper function to generate random id's for our todo's
const generateRandomId = () => {
  return Math.floor(Math.random() * 10000 + 1);
};

// Our todo Schema
export const todoSchema = z.object({
  id: z.number().default(generateRandomId),
  title: z.string()
  completed: z.boolean().default(false),
});

> // Create a new Todo type using our todo schema

Reviewer 2: Maybe explain what z.infer is doing?

![Testing our APIs: adding todos](/articles/getting-started-with-trpc/testing-our-apis-adding-todos.png)

Now, let's list them using our `todo.all` API

![Testing our APIs: displaying todos](/articles/getting-started-with-trpc/testing-our-apis-displaying-all-todos.png)

Now, let's update one of the todo's

![Testing our APIs: updating todos](/articles/getting-started-with-trpc/testing-our-apis-updating-todos.png)

And finally, let's delete one of them

![Testing our APIs: deleting todos](/articles/getting-started-with-trpc/testing-our-apis-deleting-todos.png)

Tada 🥳! Everything is working as it should!

## Conclusion

In this article, we looked at what tRPC is by building a simple todo API. tRPC is a library that allows you to set up end-to-end typing, which leads to fewer bugs and a better development experience! tRPC can be used in conjunction with many front-end frameworks such as React, React Native, Vue, and even Svelte! For more information, take a look at the [tRPC docs](https://trpc.io/docs/v10/). The code used in this article can be found in this GitHub repo.

> In the second part of this two-part series, we will be showcasing the power of having end-to-end typing by consuming our newly created todo API!

Reviewer 2: In the front-end?

---

> title: 'Getting started with tRPC v10 by building a todo app

Reviewer: Missing a closing quote
